File: jung-api/src/main/java/edu/uci/ics/jung/graph/ObservableNetwork.java
package edu.uci.ics.jung.graph;

import static com.google.common.base.Preconditions.checkNotNull;
import static java.util.Collections.synchronizedList;

import com.google.common.collect.ImmutableList;
import com.google.common.graph.ElementOrder;
import com.google.common.graph.EndpointPair;
import com.google.common.graph.Graph;
import com.google.common.graph.MutableNetwork;
import edu.uci.ics.jung.graph.event.NetworkEvent;
import edu.uci.ics.jung.graph.event.NetworkEventListener;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.function.Supplier;

/**
 * A decorator class for graphs which generates events
 *
 * @author Joshua O'Madadhain
 */
// TODO: Add tests in similar fashion to CTreeTest and AbstractCTreeTest
public class ObservableNetwork<N, E> implements MutableNetwork<N, E> {

  List<NetworkEventListener<N, E>> listenerList =
      synchronizedList(new ArrayList<NetworkEventListener<N, E>>());
  MutableNetwork<N, E> delegate;

  /**
   * Creates a new instance based on the provided {@code delegate}.
   *
   * @param delegate the graph on which this class operates
   */
  public ObservableNetwork(MutableNetwork<N, E> delegate) {
    this.delegate = checkNotNull(delegate, "delegate");
  }

  /**
   * Adds {@code l} as a listener to this graph.
   *
   * @param l the listener to add
   */
  public void addGraphEventListener(NetworkEventListener<N, E> l) {
    listenerList.add(checkNotNull(l, "l"));
  }

  /**
   * Removes {@code l} as a listener to this graph.
   *
   * @param l the listener to remove
   */
  public void removeGraphEventListener(NetworkEventListener<N, E> l) {
    listenerList.remove(checkNotNull(l, "l"));
  }

  protected void fireGraphEvent(Supplier<NetworkEvent<N, E>> evtSupplier) {
    checkNotNull(evtSupplier, "evtSupplier");
    if (listenerList.isEmpty()) return;

    NetworkEvent<N, E> evt = evtSupplier.get();
    checkNotNull(evt, "evt");

    for (NetworkEventListener<N, E> listener : listenerList) {
      listener.handleGraphEvent(evt);
    }
  }

  @Override
  public boolean addEdge(N v1, N v2, E e) {
    boolean state = delegate.addEdge(v1, v2, e);
    if (state) {
      fireGraphEvent(() -> new NetworkEvent.Edge<>(delegate, NetworkEvent.Type.EDGE_ADDED, e));
    }
    return state;
  }

  @Override
  public boolean addNode(N node) {
    boolean state = delegate.addNode(node);
    if (state) {
      fireGraphEvent(() -> new NetworkEvent.Node<>(delegate, NetworkEvent.Type.NODE_ADDED, node));
    }
    return state;
  }

  @Override
  public boolean removeEdge(E edge) {
    boolean state = delegate.removeEdge(edge);
    if (state) {
      fireGraphEvent(
          () -> new NetworkEvent.Edge<N, E>(delegate, NetworkEvent.Type.EDGE_REMOVED, (E) edge));
    }
    return state;
  }

  @Override
  public boolean removeNode(N node) {
    // remove all incident edges first, so that the appropriate events will
    // be fired (otherwise they'll be removed inside {@code
    // delegate.removeNode}
    // and the events will not be fired)
    List<E> incident_edges = ImmutableList.copyOf(delegate.incidentEdges(node));
    for (E e : incident_edges) {
      this.removeEdge(e);
    }

    boolean state = delegate.removeNode(node);
    if (state) {
      fireGraphEvent(
          () -> new NetworkEvent.Node<>(delegate, NetworkEvent.Type.NODE_REMOVED, (N) node));
    }
    return state;
  }

  @Override
  public Set<E> adjacentEdges(E edge) {
    return delegate.adjacentEdges(edge);
  }

  @Override
  public Set<N> adjacentNodes(N node) {
    return delegate.adjacentNodes(node);
  }

  @Override
  public boolean allowsParallelEdges() {
    return delegate.allowsParallelEdges();
  }

  @Override
  public boolean allowsSelfLoops() {
    return delegate.allowsSelfLoops();
  }

  @Override
  public Graph<N> asGraph() {
    return delegate.asGraph();
  }

  @Override
  public int degree(N node) {
    return delegate.degree(node);
  }

  @Override
  public ElementOrder<E> edgeOrder() {
    return delegate.edgeOrder();
  }

  @Override
  public Set<E> edges() {
    return delegate.edges();
  }

  @Override
  public Set<E> edgesConnecting(N nodeU, N nodeV) {
    return delegate.edgesConnecting(nodeU, nodeV);
  }

  @Override
  public Optional<E> edgeConnecting(N nodeU, N nodeV) {
    return delegate.edgeConnecting(nodeU, nodeV);
  }

  @Override
  public E edgeConnectingOrNull(N nodeU, N nodeV) {
    return delegate.edgeConnectingOrNull(nodeU, nodeV);
  }

  @Override
  public boolean hasEdgeConnecting(N nodeU, N nodeV) {
    return delegate.hasEdgeConnecting(nodeU, nodeV);
  }

  @Override
  public int inDegree(N node) {
    return delegate.inDegree(node);
  }

  @Override
  public Set<E> inEdges(N node) {
    return delegate.inEdges(node);
  }

  @Override
  public Set<E> incidentEdges(N node) {
    return delegate.incidentEdges(node);
  }

  @Override
  public EndpointPair<N> incidentNodes(E edge) {
    return delegate.incidentNodes(edge);
  }

  @Override
  public boolean isDirected() {
    return delegate.isDirected();
  }

  @Override
  public ElementOrder<N> nodeOrder() {
    return delegate.nodeOrder();
  }

  @Override
  public Set<N> nodes() {
    return delegate.nodes();
  }

  @Override
  public int outDegree(N node) {
    return delegate.outDegree(node);
  }

  @Override
  public Set<E> outEdges(N node) {
    return delegate.outEdges(node);
  }

  @Override
  public Set<N> predecessors(N node) {
    return delegate.predecessors(node);
  }

  @Override
  public Set<N> successors(N node) {
    return delegate.successors(node);
  }
}

File: jung-api/src/main/kotlin/edu/uci/ics/jung/graph/ObservableNetwork.kt
package edu.uci.ics.jung.graph

import com.google.common.base.Preconditions.checkNotNull
import com.google.common.collect.ImmutableList
import com.google.common.graph.ElementOrder
import com.google.common.graph.EndpointPair
import com.google.common.graph.Graph
import com.google.common.graph.MutableNetwork
import edu.uci.ics.jung.graph.event.NetworkEvent
import edu.uci.ics.jung.graph.event.NetworkEventListener
import java.util.Collections.synchronizedList
import java.util.Optional
import java.util.function.Supplier

/**
 * A decorator class for graphs which generates events
 *
 * @author Joshua O'Madadhain
 */
// TODO: Add tests in similar fashion to CTreeTest and AbstractCTreeTest
class ObservableNetwork<N : Any, E : Any>(
  private val delegate: MutableNetwork<N, E>
) : MutableNetwork<N, E> {

  private val listenerList: MutableList<NetworkEventListener<N, E>> =
    synchronizedList(ArrayList())

  init {
    checkNotNull(delegate, "delegate")
  }

  /**
   * Adds [l] as a listener to this graph.
   *
   * @param l the listener to add
   */
  fun addGraphEventListener(l: NetworkEventListener<N, E>) {
    listenerList.add(checkNotNull(l, "l"))
  }

  /**
   * Removes [l] as a listener to this graph.
   *
   * @param l the listener to remove
   */
  fun removeGraphEventListener(l: NetworkEventListener<N, E>) {
    listenerList.remove(checkNotNull(l, "l"))
  }

  protected fun fireGraphEvent(evtSupplier: Supplier<NetworkEvent<N, E>>) {
    checkNotNull(evtSupplier, "evtSupplier")
    if (listenerList.isEmpty()) return

    val evt = checkNotNull(evtSupplier.get(), "evt")

    for (listener in listenerList) {
      listener.handleGraphEvent(evt)
    }
  }

  override fun addEdge(v1: N, v2: N, e: E): Boolean {
    val state = delegate.addEdge(v1, v2, e)
    if (state) {
      fireGraphEvent(Supplier { NetworkEvent.Edge(delegate, NetworkEvent.Type.EDGE_ADDED, e) })
    }
    return state
  }

  override fun addEdge(endpoints: EndpointPair<N>, edge: E): Boolean {
    val state = delegate.addEdge(endpoints, edge)
    if (state) {
      fireGraphEvent(Supplier { NetworkEvent.Edge(delegate, NetworkEvent.Type.EDGE_ADDED, edge) })
    }
    return state
  }

  override fun addNode(node: N): Boolean {
    val state = delegate.addNode(node)
    if (state) {
      fireGraphEvent(Supplier { NetworkEvent.Node(delegate, NetworkEvent.Type.NODE_ADDED, node) })
    }
    return state
  }

  override fun removeEdge(edge: E): Boolean {
    val state = delegate.removeEdge(edge)
    if (state) {
      fireGraphEvent(Supplier { NetworkEvent.Edge(delegate, NetworkEvent.Type.EDGE_REMOVED, edge) })
    }
    return state
  }

  override fun removeNode(node: N): Boolean {
    // remove all incident edges first, so that the appropriate events will
    // be fired (otherwise they'll be removed inside delegate.removeNode
    // and the events will not be fired)
    val incidentEdges = ImmutableList.copyOf(delegate.incidentEdges(node))
    for (e in incidentEdges) {
      this.removeEdge(e)
    }

    val state = delegate.removeNode(node)
    if (state) {
      fireGraphEvent(Supplier { NetworkEvent.Node(delegate, NetworkEvent.Type.NODE_REMOVED, node) })
    }
    return state
  }

  override fun adjacentEdges(edge: E): Set<E> = delegate.adjacentEdges(edge)
  override fun adjacentNodes(node: N): Set<N> = delegate.adjacentNodes(node)
  override fun allowsParallelEdges(): Boolean = delegate.allowsParallelEdges()
  override fun allowsSelfLoops(): Boolean = delegate.allowsSelfLoops()
  override fun asGraph(): Graph<N> = delegate.asGraph()
  override fun degree(node: N): Int = delegate.degree(node)
  override fun edgeOrder(): ElementOrder<E> = delegate.edgeOrder()
  override fun edges(): Set<E> = delegate.edges()
  override fun edgesConnecting(nodeU: N, nodeV: N): Set<E> = delegate.edgesConnecting(nodeU, nodeV)
  override fun edgesConnecting(endpoints: EndpointPair<N>): Set<E> = delegate.edgesConnecting(endpoints)
  override fun edgeConnecting(nodeU: N, nodeV: N): Optional<E> = delegate.edgeConnecting(nodeU, nodeV)
  override fun edgeConnecting(endpoints: EndpointPair<N>): Optional<E> = delegate.edgeConnecting(endpoints)
  override fun edgeConnectingOrNull(nodeU: N, nodeV: N): E? = delegate.edgeConnectingOrNull(nodeU, nodeV)
  override fun edgeConnectingOrNull(endpoints: EndpointPair<N>): E? = delegate.edgeConnectingOrNull(endpoints)
  override fun hasEdgeConnecting(nodeU: N, nodeV: N): Boolean = delegate.hasEdgeConnecting(nodeU, nodeV)
  override fun hasEdgeConnecting(endpoints: EndpointPair<N>): Boolean = delegate.hasEdgeConnecting(endpoints)
  override fun inDegree(node: N): Int = delegate.inDegree(node)
  override fun inEdges(node: N): Set<E> = delegate.inEdges(node)
  override fun incidentEdges(node: N): Set<E> = delegate.incidentEdges(node)
  override fun incidentNodes(edge: E): EndpointPair<N> = delegate.incidentNodes(edge)
  override fun isDirected(): Boolean = delegate.isDirected()
  override fun nodeOrder(): ElementOrder<N> = delegate.nodeOrder()
  override fun nodes(): Set<N> = delegate.nodes()
  override fun outDegree(node: N): Int = delegate.outDegree(node)
  override fun outEdges(node: N): Set<E> = delegate.outEdges(node)
  override fun predecessors(node: N): Set<N> = delegate.predecessors(node)
  override fun successors(node: N): Set<N> = delegate.successors(node)
}
